@afc163 @benjycui 

求助，我再提交代码进行代码检查时，检查失败，信息如下：
```
antd-tools run compile
components/input/Input.tsx(202,24): error TS2531: Object is possibly 'null'.
components/input/Input.tsx(202,41): error TS2339: Property 'paddingLeft' does not exist on type 'never'.
components/input/Input.tsx(203,25): error TS2531: Object is possibly 'null'.
components/input/Input.tsx(203,42): error TS2339: Property 'paddingRight' does not exist on type 'never'.
[09:48:38] TypeScript: 4 semantic errors
[09:48:38] TypeScript: emit succeeded (with errors)
```


相关代码如下：
```jsx
// ...........省略部分代码
state = {
    textareaStyles: null,
    inputStyles: null,
    inputStylesCache: null,
};

// ...........省略部分代码
resizeInputPadding = () => {
    const { type, prefix, suffix } = this.props;
    let { inputStylesCache } = this.state;

    if (type === 'textarea' || (!prefix && !suffix && !inputStylesCache)) {
      return;
    }

    if (inputStylesCache === null) {
      const inputStyles = computedStyle(this.refs.input);
      inputStylesCache = assign({}, {
        paddingLeft: parseFloat(inputStyles.paddingLeft),
        paddingRight: parseFloat(inputStyles.paddingRight),
      });
    }

    if (!prefix && !suffix) {
      this.setState({
        inputStyles: null,
        inputStylesCache: null,
      });
    } else {
      const prefixWidth = this.calcPrefixOrSuffixWidth(this.refs.prefix);
      const suffixWidth = this.calcPrefixOrSuffixWidth(this.refs.suffix);

      this.setState({
        inputStyles: assign({}, {
          // 注释：出错代码在下面👇👇👇👇
          paddingLeft: inputStylesCache.paddingLeft + prefixWidth,
          paddingRight: inputStylesCache.paddingRight + suffixWidth,
        }),
        inputStylesCache,
      });
    }
  }
// ...........省略部分代码
```

失败位置代码截图如下：
![image](https://cloud.githubusercontent.com/assets/2840462/21412710/42a23634-c82d-11e6-856f-99a7342ed9ed.png)

--- a/components/input/Input.tsx
+++ b/components/input/Input.tsx
@@ -53,6 +53,7 @@ export interface InputProps {
   autosize?: boolean | AutoSizeType;
   autoComplete?: 'on' | 'off';
   style?: React.CSSProperties;
+  prefix?: React.ReactNode;
   suffix?: React.ReactNode;
 }
 
@@ -83,6 +84,7 @@ export default class Input extends Component<InputProps, any> {
     autosize: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
     onPressEnter: PropTypes.func,
     onKeyDown: PropTypes.func,
+    prefix: PropTypes.node,
     suffix: PropTypes.node,
   };
 
@@ -144,7 +146,7 @@ export default class Input extends Component<InputProps, any> {
     this.refs.input.focus();
   }
 
-  renderLabledInput(children) {
+  renderLabeledInput(children) {
     const props = this.props;
 
     // Not wrap when there is not addons
@@ -180,6 +182,34 @@ export default class Input extends Component<InputProps, any> {
     );
   }
 
+  renderLabeledIcon(children) {
+    const { props } = this;
+
+    if (props.type === 'textarea' || (!props.prefix && !props.suffix)) {
+      return children;
+    }
+
+    const prefix = props.prefix ? (
+      <span className={`${props.prefixCls}-prefix`}>
+        {props.prefix}
+      </span>
+    ) : null;
+
+    const suffix = props.suffix ? (
+      <span className={`${props.prefixCls}-suffix`}>
+        {props.suffix}
+      </span>
+    ) : null;
+
+    return (
+      <span className={`${props.prefixCls}-preSuffix-wrapper`}>
+        {prefix}
+        {children}
+        {suffix}
+      </span>
+    );
+  }
+
   renderInput() {
     const props = assign({}, this.props);
     // Fix https://fb.me/react-unknown-prop
@@ -189,6 +219,7 @@ export default class Input extends Component<InputProps, any> {
       'autosize',
       'addonBefore',
       'addonAfter',
+      'prefix',
       'suffix',
     ]);
 
@@ -222,8 +253,7 @@ export default class Input extends Component<InputProps, any> {
           />
         );
       default:
-        const withSuffix = props.suffix && !props.textarea;
-        const input = (
+        return this.renderLabeledIcon(
           <input
             {...otherProps}
             className={inputClassName}
@@ -231,19 +261,10 @@ export default class Input extends Component<InputProps, any> {
             ref="input"
           />
         );
-
-        return withSuffix ? (
-          <span className={`${prefixCls}-suffix-wrapper`}>
-            {input}
-            <span className={`${prefixCls}-suffix`}>
-              {props.suffix}
-            </span>
-          </span>
-        ) : input;
     }
   }
 
   render() {
-    return this.renderLabledInput(this.renderInput());
+    return this.renderLabeledInput(this.renderInput());
   }
 }

--- a/components/input/style/index.less
+++ b/components/input/style/index.less
@@ -14,6 +14,6 @@
 }
 
 // Input with suffix
-.@{ant-prefix}-input-suffix-wrapper {
-  .input-suffix-wrapper(~"@{ant-prefix}-input");
+.@{ant-prefix}-input-preSuffix-wrapper {
+  .input-preSuffix-wrapper(~"@{ant-prefix}-input");
 }

--- a/components/input/style/mixin.less
+++ b/components/input/style/mixin.less
@@ -1,6 +1,8 @@
 @import "../../style/themes/default";
 @import "../../style/mixins/index";
 
+@input-presuffix-width: 16;
+
 // size mixins for input
 .input-lg() {
   padding: @input-padding-vertical-lg @input-padding-horizontal;
@@ -207,14 +209,15 @@
     }
   }
 
-  > .@{inputClass}-suffix-wrapper {
+  > .@{inputClass}-preSuffix-wrapper {
     &:not(:first-child) .@{inputClass} {
       border-bottom-left-radius: 0;
       border-top-left-radius: 0;
     }
 
-    &:last-child .@{inputClass} {
-      border-radius: 0 @border-radius-base @border-radius-base 0;
+    &:not(:last-child) .@{inputClass} {
+      border-bottom-right-radius: 0;
+      border-top-right-radius: 0;
     }
   }
 
@@ -248,34 +251,51 @@
     .input-sm();
   }
 
-  .@{inputClass}-suffix-wrapper {
+  .@{inputClass}-preSuffix-wrapper {
     display: table-cell;
     width: 100%;
     float: left;
   }
 }
 
-.input-suffix-wrapper(@inputClass) {
+.input-preSuffix-wrapper(@inputClass) {
   position: relative;
   display: inline-block;
   width: 100%;
 
+  .@{inputClass} {
+    z-index: 1;
+  }
+
+  .@{inputClass}-prefix,
   .@{inputClass}-suffix {
     position: absolute;
     top: 0;
-    right: @input-padding-horizontal;
-    height: @input-height-base;
-    line-height: @input-height-base;
+    bottom: 0;
     z-index: 2;
+
+    &:after {
+      content: '';
+      display: inline-block;
+      width: 0;
+      height: 100%;
+      vertical-align: middle;
+    }
   }
 
-  .@{inputClass}-lg + .@{inputClass}-suffix {
-    height: @input-height-lg;
-    line-height: @input-height-lg;
+  .@{inputClass}-prefix {
+    left: @input-padding-horizontal;
   }
 
-  .@{inputClass}-sm + .@{inputClass}-suffix {
-    height: @input-height-sm;
-    line-height: @input-height-sm;
+  .@{inputClass}-suffix {
+    right: @input-padding-horizontal;
+  }
+
+  .@{inputClass}:not(:first-child) {
+    padding-left: @input-padding-horizontal + @input-presuffix-width;
+  }
+
+  .@{inputClass}:not(:last-child) {
+    padding-right: @input-padding-horizontal + @input-presuffix-width;
   }
 }

